Public/Src/Cache/ContentStore/Library/Utils/BandwidthChecker.cs

```diff
@@ -45,93 +45,76 @@ public async Task<CopyFileResult> CheckBandwidthAtIntervalAsync(OperationContext
+                var (result, bytesCopied) = await impl();
+                timer.Stop();
 
-                try
-                {
-                    return await impl();
-                }
-                finally
-                {
-                    timer.Stop();
-                    var endPosition = destinationStream.Position;
+                // Bandwidth checker expects speed in MiB/s, so convert it.
+                var speed = bytesCopied / timer.Elapsed.TotalSeconds / BytesInMb;
+                _historicalBandwidthLimitSource.AddBandwidthRecord(speed);
 
-                    // Bandwidth checker expects speed in MiB/s, so convert it.
-                    var bytesCopied = endPosition - startPosition;
-                    var speed = bytesCopied / timer.Elapsed.TotalSeconds / (1024 * 1024);
-                    _historicalBandwidthLimitSource.AddBandwidthRecord(speed);
-                }
+                return result;
             }
             else
             {
-                return await impl();
+                return (await impl()).result;
             }
 
-            async Task<CopyFileResult> impl()
+            async Task<(CopyFileResult result, long bytesCopied)> impl()
             {
                 // This method should not fail with exceptions because the resulting task may be left unobserved causing an application to crash
                 // (given that the app is configured to fail on unobserved task exceptions).
                 var minimumSpeedInMbPerSec = _bandwidthLimitSource.GetMinimumSpeedInMbPerSec() * _config.BandwidthLimitMultiplier;
                 minimumSpeedInMbPerSec = Math.Min(minimumSpeedInMbPerSec, _config.MaxBandwidthLimit);
 
-                long previousPosition = 0;
+                var startPosition = tryGetPosition(destinationStream, out var pos) ? pos : 0;
+                long previousPosition = startPosition;
                 var copyCompleted = false;
                 using var copyCancellation = CancellationTokenSource.CreateLinkedTokenSource(context.Token);
                 var copyTask = copyTaskFactory(copyCancellation.Token);
 
-                try
+                while (!copyCompleted)
                 {
-                    while (!copyCompleted)
+                    // Wait some time for bytes to be copied
+                    var firstCompletedTask = await Task.WhenAny(copyTask,
+                        Task.Delay(_config.BandwidthCheckInterval, context.Token));
+
+                    copyCompleted = firstCompletedTask == copyTask;
+                    if (copyCompleted)
                     {
-                        // Wait some time for bytes to be copied
-                        var firstCompletedTask = await Task.WhenAny(copyTask,
-                            Task.Delay(_config.BandwidthCheckInterval, context.Token));
+                        var result = await copyTask;
+                        var bytesCopied = result.Size ?? previousPosition - startPosition;

```

Comment:
one very minor nit: could you please add braces?
because I have no idea if this code is correct.